reference/src/layout/scalars.md

```diff
@@ -37,6 +37,13 @@ The `isize` and `usize` types are pointer-sized signed and unsigned integers.
 They have the same layout as the [pointer types] for which the pointee is
 `Sized`, and are layout compatible with C's `uintptr_t` and `intptr_t` types.
 
+> **Note**: C99 [7.18.2.4](https://port70.net/~nsz/c/c99/n1256.html#7.18.2.4)
+> requires `uintptr_t` and `intptr_t` to be at least 16-bit wide. That is, on
+> all platforms that we currently support, which have a C platform,
+> `isize`/`usize` are at least 16-bit wide. `libcore` relies on this guarantee
+> to unconditionally provide impls of `From<i16>`/`From<u16>` for
```

Comment:
`libcore` doesn't really rely on this, it just effectively has `#[cfg(all_platforms_supported_right_now)]` on the impl.
If we didn't support 16-bit platforms yet, `From<i32>` would also be unconditional.
`cfg`-s can be added on every new platform support if necessary.

I don't understand why this guarantee is useful, other libraries are free to support any platforms they want, I'm pretty sure many of them are not tested on 16-bit either and therefore not expected to work without fixes.

Reply:
First, we do agree. 

> libcore doesn't really rely on this, it just effectively has #[cfg(all_platforms_supported_right_now)] on the impl.
If we didn't support 16-bit platforms yet, From<i32> would also be unconditional.
cfg-s can be added on every new platform support if necessary.

The main argument for adding that API was that C guarantees that `uintptr_t`/`intptr_t` are at least 16-bit wide (the source code even calls it as such). I mentioned [here](https://github.com/rust-lang/unsafe-code-guidelines/pull/154#issuecomment-507142684), that there is no technical reason we couldn't cfg them out if we ever added support for an 8-bit platform: that wouldn't break any code there, because no code there exists. Still, all non-portable APIs in libcore are in `core::arch`, and this one isn't. Whether this should be available for all Rust programs or not, is kind of scope for the UCGs (more of a T-libs/T-lang/T-compiler issue). 

> I don't understand why this guarantee is useful, other libraries are free to support any platforms they want, I'm pretty sure many of them are not tested on 16-bit either and therefore not expected to work without fixes.

Note that this PR does not guarantee anything. It documents something that is true for all platforms that Rust supports today, and explicitly calls it as such. 

Why would guaranteeing this be useful ? It would make this code "`if size_of::<usize>() < 2 { ... }`" dead for all Rust programs, removing one corner case / bit-of-trivia that programmers have to remember, or that can be used to accuse code of "not being portable". I do see some value in this. 

Are we already guaranteeing this ? I don't know, I don't think so. [In previous discussions](https://github.com/rust-lang/unsafe-code-guidelines/issues/9), it appeared to be common folklore that there are some platforms that we never intend to support, like platforms where `CHAR_BITS != 8`, but 8-bit platforms were not mentioned there. 

This is why the wording is so "generic". `usize`/`isize` correspond to `uintptr_t`/`intptr_t` of the C platform, iff a C platform exists. A standard compliant C platform cannot exist for 8-bit architectures, so there we can do something else, as long as we don't break targets where a C platform exists. Right now, that something else is unspecified.

Reply:
I have been told that once the [portability lint](https://github.com/rust-lang/rust/issues/41619) feature gets farther, then we'll be in a much better position to make "potentially not portable" code in core and std. One of the exact examples given is that, with the portability lint ability in place, we'd be able to add From/Into for usize/isize/u32/i32/u64/i64 and they get cfg'd in based on your platform width, and so on.

Of course, it's had basically no progress since 2017 it looks like, so rip.

Reply:
Yeah, there's plausible ways for libcore to be ported to a new platform and lose the impl there, so let's not say it "relies" on this. Just noting the relation to C is good enough for now.

Reply:
I've reworded this a bit - is it better now?

Reply:
LGTM